deepmatch/models/comirec.py

```diff
@@ -0,0 +1,165 @@
+            history_feature_columns.append(fc)
+        else:
+            sparse_varlen_feature_columns.append(fc)
+    seq_max_len = history_feature_columns[0].maxlen
+    inputs_list = list(features.values())
+
+    embedding_matrix_dict = create_embedding_matrix(user_feature_columns + item_feature_columns, l2_reg_embedding,
+                                                    seed=seed, prefix="")
+
+    item_features = build_input_features(item_feature_columns)
+
+    query_emb_list = embedding_lookup(embedding_matrix_dict, item_features, item_feature_columns,
+                                      history_feature_list,
+                                      history_feature_list, to_list=True)
+    keys_emb_list = embedding_lookup(embedding_matrix_dict, features, history_feature_columns, history_fc_names,
+                                     history_fc_names, to_list=True)
+    dnn_input_emb_list = embedding_lookup(embedding_matrix_dict, features, sparse_feature_columns,
+                                          mask_feat_list=history_feature_list, to_list=True)
+    dense_value_list = get_dense_input(features, dense_feature_columns)
+
+    sequence_embed_dict = varlen_embedding_lookup(embedding_matrix_dict, features, sparse_varlen_feature_columns)
+    sequence_embed_list = get_varlen_pooling_list(sequence_embed_dict, features, sparse_varlen_feature_columns, to_list=True)
+
+    dnn_input_emb_list += sequence_embed_list
+
+    # keys_emb = concat_func(keys_emb_list, mask=True)
+    # query_emb = concat_func(query_emb_list, mask=True)
+
+    history_emb = PoolingLayer()(NoMask()(keys_emb_list))     # [None, MAX_LEN, emb_dim]
+    target_emb = PoolingLayer()(NoMask()(query_emb_list))
+
+    # target_emb_size = target_emb.get_shape()[-1].value
+    # max_len = history_emb.get_shape()[1].value
+    hist_len = features['hist_len']
+
+    high_capsule =None
+    if interest_extractor.lower()=='dr':
+        high_capsule = CapsuleLayer(input_units=item_embedding_dim,
+                                    out_units=item_embedding_dim, max_len=seq_max_len,
+                                    k_max=interest_num)((history_emb, hist_len))
+    elif interest_extractor.lower()=='sa':
+        history_emb_add_pos = history_emb
+        if add_pos:
+            position_embedding = PositionEncoding()(history_emb)
+            history_emb_add_pos += position_embedding # [None, MAX_LEN, emb_dim]
+
+        attn = DNN((item_embedding_dim*4, interest_num), dnn_activation, l2_reg_dnn,
+                        dnn_dropout, dnn_use_bn, output_activation=output_activation, seed=seed,
+                        name="user_dnn_attn")(history_emb_add_pos)
+        attn = tf.transpose(attn, [0, 2, 1]) # [None, interest_num, MAX_LEN]
+        seq_len_tile = tf.tile(hist_len, [1, interest_num])
+        mask = tf.sequence_mask(seq_len_tile, seq_max_len)  # [None, interest_num, MAX_LEN]
+        high_capsule = SoftmaxWeightedSum()([attn, history_emb_add_pos, mask])
```

Comment:
`SoftmaxWeightedSum` has a default parameter `dropout_rate=0.2`,do we need it here?

Reply:
We don't. I will remove it.